src/main/java/io/supertokens/passwordless/Passwordless.java

```diff
@@ -124,6 +129,121 @@ private static String generateUserInputCode() {
+                    throw new StorageTransactionLogicException(new RestartFlowException());
+                }
+                if (device.failedAttempts >= maxCodeInputAttempts) {
+                    passwordlessStorage.deleteDevice_Transaction(con, deviceIdHash);
+                    passwordlessStorage.commitTransaction(con);
+                    throw new StorageTransactionLogicException(new RestartFlowException());
+                }
+
+                PasswordlessCode code = passwordlessStorage.getCodeByLinkCodeHash_Transaction(con, linkCodeHash);
+                if (code == null || code.createdAt < System.currentTimeMillis() - passwordlessCodeLifetime) {
+                    if (deviceId != null) {
+                        if (device.failedAttempts >= maxCodeInputAttempts) {
+                            passwordlessStorage.deleteDevice_Transaction(con, deviceIdHash);
+                            passwordlessStorage.commitTransaction(con);
+                            throw new StorageTransactionLogicException(new RestartFlowException());
+                        } else {
+                            passwordlessStorage.incrementDeviceFailedAttemptCount_Transaction(con, deviceIdHash);
+                            passwordlessStorage.commitTransaction(con);
+
+                            if (code != null) {
+                                throw new StorageTransactionLogicException(new ExpiredUserInputCodeException(
+                                        device.failedAttempts + 1, maxCodeInputAttempts));
+                            } else {
+                                throw new StorageTransactionLogicException(new IncorrectUserInputCodeException(
+                                        device.failedAttempts + 1, maxCodeInputAttempts));
+                            }
+                        }
+                    }
+                    throw new StorageTransactionLogicException(new RestartFlowException());
+                }
+
+                if (device.email != null) {
+                    passwordlessStorage.deleteDevicesByEmail_Transaction(con, device.email);
+                } else if (device.phoneNumber != null) {
+                    passwordlessStorage.deleteDevicesByPhoneNumber_Transaction(con, device.phoneNumber);
+                }
+
+                passwordlessStorage.commitTransaction(con);
+                return device;
+            });
+        } catch (StorageTransactionLogicException e) {
+            if (e.actualException instanceof ExpiredUserInputCodeException) {
+                throw (ExpiredUserInputCodeException) e.actualException;
+            }
+            if (e.actualException instanceof IncorrectUserInputCodeException) {
+                throw (IncorrectUserInputCodeException) e.actualException;
+            }
+            if (e.actualException instanceof RestartFlowException) {
+                throw (RestartFlowException) e.actualException;
+            }
+            throw e;
+        }
+
```

Comment:
Add a comment here saying that if the execution comes here, it means that we have successfully consumed the code